haystack/components/preprocessors/recursive_splitter.py

```diff
@@ -0,0 +1,241 @@
+            if curr_separator == "sentence":
+                # using the custom NLTK-based sentence tokenizer
+                sentence_with_spans = self.nltk_tokenizer.split_sentences(text)
+                splits = [sentence["sentence"] for sentence in sentence_with_spans]
+            else:
+                # apply current separator regex to split text
+                escaped_separator = re.escape(curr_separator)
+                splits = re.split(escaped_separator, text)
+
+            if len(splits) == 1:  # go to next separator, if current separator not found
+                continue
+
+            chunks = []
+            current_chunk: List[str] = []
+            current_length = 0
+
+            # check splits, if any is too long, recursively chunk it, otherwise add to current chunk
+            for idx, split in enumerate(splits):
+                split_text = split
+
+                # add separator to the split, if it's not the last split
+                if self.keep_separator and curr_separator != "sentence" and idx < len(splits) - 1:
+                    split_text = split + curr_separator
+
+                # if adding this split exceeds chunk_size, process current_chunk
+                if current_length + len(split_text) > self.split_length:
+                    if current_chunk:  # keep the good splits
+                        chunks.append("".join(current_chunk))
+                        current_chunk = []
+                        current_length = 0
+
+                    # recursively handle splits that are too large
+                    if len(split_text) > self.split_length:
+                        if curr_separator == self.separators[-1]:
+                            # tried the last separator, can't split further, break the loop and fall back to
+                            # character-level chunking
+                            break
+                        chunks.extend(self._chunk_text(split_text))
+                    else:
+                        chunks.append(split_text)
+                else:
+                    current_chunk.append(split_text)
+                    current_length += len(split_text)
+
+            if current_chunk:
+                chunks.append("".join(current_chunk))
+
+            if self.split_overlap > 0:
+                chunks = self._apply_overlap(chunks)
+
+            if chunks:
+                return chunks
+
+        # if no separator worked, fall back to character-level chunking
+        return [text[i : i + self.split_length] for i in range(0, len(text), self.split_length - self.split_overlap)]
+
+    def _run_one(self, doc: Document) -> List[Document]:
+        new_docs: List[Document] = []
+        chunks = self._chunk_text(doc.content)  # type: ignore # the caller already check for a non-empty doc.content
+        chunks = chunks[:-1] if len(chunks[-1]) == 0 else chunks  # remove last empty chunk
```

Comment:
Out of curiosity in what situations do we end up with an empty chunk at the end?

Reply:
When the splitting character comes at the end, for instance:

```python
import re
text = "This is a test. Another sentence. And one more."
re.split("(\\.)",text)
```

results in:

```python
['This is a test', '.', ' Another sentence', '.', ' And one more', '.', '']
```

But the removal of `keep_separator: bool` introduces a fix that solves this

Reply:
ahh okay thanks for the explanation!